cpp/src/arrow/flight/sql/odbc/odbcabstraction/odbc_impl/odbc_statement.cc

```diff
@@ -663,9 +663,12 @@ void ODBCStatement::SetStmtAttr(SQLINTEGER statementAttribute, SQLPOINTER value,
       throw DriverException("Invalid attribute: " + std::to_string(attributeToWrite),
                             "HY092");
   }
-  if (!successfully_written) {
+  if (successfully_written) {
+    return SQL_SUCCESS;
+  } else {
     GetDiagnostics().AddWarning("Optional value changed.", "01S02",
                                 driver::odbcabstraction::ODBCErrorCodes_GENERAL_WARNING);
+    return SQL_SUCCESS_WITH_INFO;
   }
```

Comment:
I am not sure if we need to change `SetStmtAttr` to return `SQLRETURN`, because [`odbc_handle.h`](https://github.com/apache/arrow/blob/73472a958d50ffb181e9daa0ac891d7e2a76a637/cpp/src/arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/odbc_impl/odbc_handle.h#L64) should catch any warnings and return `SQL_SUCCESS_WITH_INFO`